src/tests/test_dilithium.cpp

```diff
@@ -0,0 +1,375 @@
+      {
+      std::vector<Test::Result> results;
+
+#if defined(BOTAN_HAS_DILITHIUM)
+      results.push_back(run_roundtrip("Dilithium_4x4_Common",
+                                      Botan::DilithiumDimension::Dilithium4x4,
+                                      Botan::DilithiumFlavor::Deterministic));
+      results.push_back(run_roundtrip("Dilithium_6x5_Common",
+                                      Botan::DilithiumDimension::Dilithium6x5,
+                                      Botan::DilithiumFlavor::Deterministic));
+      results.push_back(run_roundtrip("Dilithium_8x7_Common",
+                                      Botan::DilithiumDimension::Dilithium8x7,
+                                      Botan::DilithiumFlavor::Deterministic));
+      results.push_back(run_roundtrip("Dilithium_4x4_Randomized",
+                                      Botan::DilithiumDimension::Dilithium4x4,
+                                      Botan::DilithiumFlavor::Randomized));
+      results.push_back(run_roundtrip("Dilithium_6x5_Randomized",
+                                      Botan::DilithiumDimension::Dilithium6x5,
+                                      Botan::DilithiumFlavor::Randomized));
+      results.push_back(run_roundtrip("Dilithium_8x7_Randomized",
+                                      Botan::DilithiumDimension::Dilithium8x7,
+                                      Botan::DilithiumFlavor::Randomized));
+#endif
+
+#if defined(BOTAN_HAS_DILITHIUM_AES)
+      results.push_back(run_roundtrip("Dilithium_4x4_AES",
+                                      Botan::DilithiumDimension::Dilithium4x4,
+                                      Botan::DilithiumFlavor::AesDeterministic));
+      results.push_back(run_roundtrip("Dilithium_6x5_AES",
+                                      Botan::DilithiumDimension::Dilithium6x5,
+                                      Botan::DilithiumFlavor::AesDeterministic));
+      results.push_back(run_roundtrip("Dilithium_8x7_AES",
+                                      Botan::DilithiumDimension::Dilithium8x7,
+                                      Botan::DilithiumFlavor::AesDeterministic));
+      results.push_back(run_roundtrip("Dilithium_4x4_AESRandomized",
+                                      Botan::DilithiumDimension::Dilithium4x4,
+                                      Botan::DilithiumFlavor::AesRandomized));
+      results.push_back(run_roundtrip("Dilithium_6x5_AESRandomized",
+                                      Botan::DilithiumDimension::Dilithium6x5,
+                                      Botan::DilithiumFlavor::AesRandomized));
+      results.push_back(run_roundtrip("Dilithium_8x7_AESRandomized",
+                                      Botan::DilithiumDimension::Dilithium8x7,
+                                      Botan::DilithiumFlavor::AesRandomized));
+#endif
+
+         return results;
+      }
+   };
+
+BOTAN_REGISTER_TEST("dilithium", "dilithium_roundtrips", DilithiumRoundtripTests);
+
```

Comment:
This is missing a key generation test (derived from `PK_Key_Generation_Test`) which for all available parameters, creates a random key via the generic interface, then encodes/decodes it. This is somewhat redundant with the existing round trip tests, except that it also exercises the algorithm agnostic codepaths used by the CLI, X.509, etc.

Reply:
Very helpful pointer, thanks. `class DilithiumMode` is now pretty much equivalent to `class KyberMode`. Also, the generic test found a bug: `m_nist_strength` was set to "2", "3" and "5", rather than the well-known bit-values. 😨